File: Top Interview Questions/155. Min Stack.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.m = float('inf')
        self.stack = []

        
    def push(self, x: int) -> None:
        if len(self.stack) == 0:
            self.stack.append(0)
            self.m = x
        else:
            self.stack.append(x - self.m)
            self.m = min(self.m, x)

            
    def pop(self) -> None:
        if len(self.stack) > 0:
            p = self.stack.pop()
            self.m -= p if p < 0 else 0
        
        
    def top(self) -> int:
        t = self.stack[-1]
        return (t + self.m if t > 0 else self.m)
        

    def getMin(self) -> int:
        return self.m
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Top Interview Questions/Design/155. Min Stack.java
class MinStack {

    /** initialize your data structure here. */
    private long min;
    private Stack<Long> stack;
    
    public MinStack() {
        stack = new Stack<>();
    }
    
    public void push(int x) {
        if (stack.isEmpty()){
            min = x;
            stack.push(0L);
        }else{
            stack.push(x - min);
            if (x < min){
                min = x;
            }
        }
    }
    
    public void pop() {
        if (!stack.isEmpty()){
            long top = stack.pop();
            if (top < 0){
                min = min - top;
            }
        }else{
            System.out.println("Error");
        }
        
    }
    
    public int top() {
        if (stack.peek() > 0){
            return (int)(stack.peek() + min);
        }else{
            return (int)min;
        }
    }
    
    public int getMin() {
        return (int)min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */